stan/math/prim/fun/gp_dot_prod_cov.hpp

```diff
@@ -54,15 +54,21 @@ gp_dot_prod_cov(const std::vector<Eigen::Matrix<T_x, Eigen::Dynamic, 1>> &x,
   }
 
   T_sigma sigma_sq = square(sigma);
-
-  for (size_t i = 0; i < (x_size - 1); ++i) {
-    cov(i, i) = sigma_sq + dot_self(x[i]);
-    for (size_t j = i + 1; j < x_size; ++j) {
-      cov(i, j) = sigma_sq + dot_product(x[i], x[j]);
-      cov(j, i) = cov(i, j);
+  size_t block_size = 10;
+
+  for (size_t jb = 0; jb < x_size; jb += block_size) {
+    for (size_t ib = jb; ib < x_size; ib += block_size) {
+      for (size_t j = jb; j < std::min(x_size, jb + block_size); ++j) {
+        cov.coeffRef(j, j) = sigma_sq + dot_self(x[j]);
+        for (size_t i = std::max(ib, j + 1);
+             i < std::min(x_size, ib + block_size); ++i) {
+          cov.coeffRef(j, i) = cov.coeffRef(i, j)
+              = sigma_sq + dot_product(x[i], x[j]);
+        }
+      }
     }
   }
```

Comment:
How does the block schema work here wrt cache friendlyness? I'm surprised most of the time isn't being spent in `dot_product(x[i], x[j])`. I wonder how much going off of fixed block sizes is actually helping compiler loop optimizations.

Reply:
> I'm surprised most of the time isn't being spent in dot_product(x[i], x[j])

Virtually all time is spent in dot product. That is why it is faster if the data is in cache. Loop optimizations probably suffer, but less transfers outweighs extra loops.

Reply:
What if we just had an if that for x_size < block_size did the original and for > did the above version?

Reply:
While I could add that, I think the extra code is not worth it just for ~5% speedup at tiny sizes. Also with the new bench ne code looks faster regardless of size.